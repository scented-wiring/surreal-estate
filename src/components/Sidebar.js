import React, { useState } from "react";
import "../styles/Sidebar.css";
import { Link, useLocation, useHistory } from "react-router-dom";
import qs from "qs";

const Sidebar = () => {
  const { search } = useLocation();
  const history = useHistory();

  const [query, setQuery] = useState("");

  const buildQueryString = (operation, valueObj) => {
    const currentQueryParams = qs.parse(search, { ignoreQueryPrefix: true });
    const newQueryParams = {
      ...currentQueryParams,
      [operation]: JSON.stringify({
        ...JSON.parse(currentQueryParams[operation] || "{}"),
        ...valueObj,
      }),
    };

    return qs.stringify(newQueryParams, {
      addQueryPrefix: true,
      encode: false,
    });
  };

  const handleSearch = (event) => {
    event.preventDefault();

    const newQueryString = buildQueryString("query", {
      title: { $regex: query, $options: "i" },
    });

    history.push(newQueryString);
  };

  return (
    <div className="sidebar">
      <form onSubmit={handleSearch}>
        <input
          className="search-input"
          type="text"
          value={query}
          id="input"
          onChange={(e) => setQuery(e.target.value)}
        />
        <button className="search-button" type="submit">
          Search
        </button>
      </form>

      <div className="item-title">Filter by city</div>
      <div className="item">
        <Link to={buildQueryString("query", { city: "Manchester" })}>
          Manchester
        </Link>
      </div>
      <div className="item">
        <Link to={buildQueryString("query", { city: "Leeds" })}>Leeds</Link>
      </div>
      <div className="item">
        <Link to={buildQueryString("query", { city: "Sheffield" })}>
          Sheffield
        </Link>
      </div>
      <div className="item">
        <Link to={buildQueryString("query", { city: "Liverpool" })}>
          Liverpool
        </Link>
      </div>

      <div className="item-title">Sort by</div>
      <div className="item">
        <Link to={buildQueryString("sort", { price: -1 })}>
          Price - Descending
        </Link>
      </div>
      <div className="item">
        <Link to={buildQueryString("sort", { price: 1 })}>
          Price - Ascending
        </Link>
      </div>
    </div>
  );
};

export default Sidebar;
