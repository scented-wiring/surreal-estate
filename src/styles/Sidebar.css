.sidebar {
  flex: 0 1 230px;
  width: 270px;
  padding-left: 20px;
  font-size: 1em;
  background-color: white;
  padding-bottom: 21.8%;
}

.search-input {
  margin: 10px 0 10px 0;
}

.search-button {
  height: 21px;
  color: white;
  background-color: purple;
}

.search-button:hover {
  background: black;
  transition: all 0.4s ease 0s;
}

.item {
  padding: 0 0 20px 0;
}

.item-title {
  padding: 0 0 20px 0;
  font-weight: bold;
  text-decoration: underline;
}
