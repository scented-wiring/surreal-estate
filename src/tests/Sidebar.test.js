import React from "react";
import { render } from "@testing-library/react";
import { MemoryRouter } from "react-router-dom";
import Sidebar from "../components/Sidebar";

describe("Sidebar", () => {
  it("renders correctly", () => {
    const { asFragment } = render(
      <MemoryRouter>
        <Sidebar />
      </MemoryRouter>
    );
    expect(asFragment).toMatchSnapshot();
  });

  it("renders a search input", () => {
    const { getByRole } = render(
      <MemoryRouter>
        <Sidebar />
      </MemoryRouter>
    );
    expect(getByRole("textbox")).toBeInTheDocument();
  });

  it("renders a search button", () => {
    const { getByRole } = render(
      <MemoryRouter>
        <Sidebar />
      </MemoryRouter>
    );
    expect(getByRole("button")).toBeInTheDocument();
  });

  it("renders two titles", () => {
    const { getAllByTestId } = render(
      <MemoryRouter>
        <Sidebar />
      </MemoryRouter>
    );
    expect(getAllByTestId("item-title")).toHaveLength(2);
  });

  it("renders six links", () => {
    const { getAllByRole } = render(
      <MemoryRouter>
        <Sidebar />
      </MemoryRouter>
    );
    expect(getAllByRole("link")).toHaveLength(6);
  });
});
