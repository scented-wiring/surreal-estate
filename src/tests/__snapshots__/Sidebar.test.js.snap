// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Sidebar renders correctly 1`] = `[Function]`;
